### PR TITLE
Reduce duplicate readlock and writelock code in MultiReaderFastList, MultiReaderHashBag and MultiReaderUnifiedSet.

### PR DESCRIPTION
**Pull request by donraab**
Signed-off-by: Donald Raab <Donald.Raab@bnymellon.com>

### CODE REVIEW

**Comment by donraab** [comment]
This was a thought experiment I wanted to run by @motlin and @nikhilnanivadekar. Feedback welcome. If this works here, then we can apply the same approach to the other two MultiReader collections to further reduce duplicate code.

**Comment by donraab** [comment]
Nice! The Scala tests failed for MultiReaderFastList testing that the appropriate readlocks and writelocks were taken for each method. I had changed some code that should have used a writelock and instead used a readlock. I added a few missing sortByPrimitive methods to the Scala test.

**Comment by donraab** [comment]
Ok build is passing now. Checkstyle and FindBugs had a couple complaints which I addressed. @motlin  and @nikhilnanivadekar let me know what you think.

**Comment by mohrezaei** [comment]
Have you looked at performance and garbage creation with these changes?

**Comment by donraab** [comment]
No, but I could do that.  I think I might be able to make some of the calls not generate instances with the lambda by using the same approach we use with `with` methods.  Just requires me writing some more code. I can do that after I return from my vacation.

**Comment by mohrezaei** [comment]
wouldn't this https://stackoverflow.com/questions/6965731/are-locks-autocloseable be a more straight forward solution? I think the wrapper could be created just once, but we'd have to change the signature of the getLock() method to return the wrapper instead.

**Comment by donraab** [comment]
How would the LockWrapper close method be implemented?  It is a ReentrantReadWriteLock.

**Comment by mohrezaei** [comment]
Two wrappers, one for read, one for write.

**Comment by donraab** [comment]
Right, figured it out after I commented.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java [hunk 1]
```diff
@@ -102,6 +103,101 @@ protected void unlockReadLock()
         this.getLock().readLock().unlock();
     }
 
+    protected <R> R applyReadLock(Function0<R> function)
```
Comment (sendilkumarn) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: Can we try using `execute-around` pattern for these methods too?

**Comment by donraab** [comment]
I'm updating all of the methods to use the new LockWrapper approach. I am changing MultiReaderFastList, MultiReaderHashBag and MultiReaderUnifiedSet to use this new pattern.

**Comment by donraab** [comment]
I refactored MultiReaderFastList, MultiReaderHashBag and MultiReaderUnifiedSet to use the try with resources approach Moh suggested.  Can you have a look @mohrezaei @motlin @nikhilnanivadekar @sendilkumarn?  I think it is working out pretty well.  I got rid of the lambda approach altogether, because it's easier to read the consistent surround approach than having multiple lambda with parameter methods in order to leverage method references.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java [hunk 2]
```diff
@@ -2576,4 +1995,52 @@ public void appendString(Appendable appendable, String start, String separator,
             return map;
         }
     }
+
+    public class ReadWriteLockWrapper
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: add static

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java [hunk 3]
```diff
@@ -2576,4 +1995,52 @@ public void appendString(Appendable appendable, String start, String separator,
...
+         * This method must be wrapped in a try block.
+         */
+        public LockWrapper acquireWriteLock()
+        {
+            this.writeLock.lock.lock();
+            return this.writeLock;
+        }
+    }
+
+    public class LockWrapper implements AutoCloseable
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: add static

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java [hunk 4]
```diff
@@ -78,231 +79,139 @@
...
 {
-    protected abstract MutableCollection<T> getDelegate();
-
-    protected abstract ReadWriteLock getLock();
-
-    protected void acquireWriteLock()
-    {
-        this.getLock().writeLock().lock();
-    }
+    protected transient ReadWriteLock lock;
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: do we still need this field?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java [hunk 5]
```diff
@@ -78,231 +79,139 @@
...
-    }
-
-    protected void acquireReadLock()
-    {
-        this.getLock().readLock().lock();
-    }
+    protected abstract MutableCollection<T> getDelegate();
 
-    protected void unlockReadLock()
+    protected ReadWriteLock getLock()
```
Comment (mohrezaei) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: do we need this method?

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java (hunk 4 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: There are other parallel classes which still depend on the field.  I will get rid of it once I get a chance to refactor those classes as well.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java (hunk 2 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: Good catch, will do.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java (hunk 5 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: I'll check for usages and remove if not used.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java (hunk 3 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: done

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java (hunk 5 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: removed

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java (hunk 2 above)
Comment (donraab) [reply eclipse-collections/src/main/java/org/eclipse/collections/impl/collection/mutable/AbstractMultiReaderMutableCollection.java]: done

**Comment by nikhilnanivadekar** [comment]
There are `finally` blocks removed from the code. Will that be ok? If it will be ok, how and when is the lock released? Sorry, if I am missing something obvious.

The build failures seem to be due to an issue with JDK download and is not related to the PR. Since the Oracle JDK builds for the failed Open JDK builds have gone through, I am fine from a build perspective.

**Comment by mohrezaei** [comment]
@nikhilnanivadekar it's [try with resources](https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

**Comment by nikhilnanivadekar** [comment]
@mohrezaei yup... it was so obvious in hindsight! :)

**Pull request merged**